src/main/scala/org/ergoplatform/nodeView/wallet/WalletScanLogic.scala

```diff
@@ -63,46 +66,91 @@ object WalletScanLogic extends ScorexLogging {
+      // todo: consider different profiles for the config, such as "user", "wallet", "app hub"
+      val expectedKeys = 20000
+      val falsePositiveRate = 0.01
+      val bf = BloomFilter.create[Array[Byte]](Funnels.byteArrayFunnel(), expectedKeys, falsePositiveRate)
+
+      registry.allUnspentBoxes().foreach { tb =>
+        bf.put(tb.box.id)
+      }
+      bf
+    }
 
+    // Resolve boxes related to mining income
     val resolvedBoxes = registry.unspentBoxes(MiningScanId).flatMap { tb =>
       val spendable = resolve(tb.box, walletVars.proverOpt, stateContext, height)
       if (spendable) Some(tb.copy(scans = Set(PaymentsScanId))) else None
     }
 
     //input tx id, input box id, tracked box
-    type InputData = Seq[(ModifierId, EncodedBoxId, TrackedBox)]
+    type InputData = Seq[(ModifierId, ModifierId, TrackedBox)]
     //outputs, input ids, related transactions
     type ScanResults = (Seq[TrackedBox], InputData, Seq[WalletTransaction])
     val initialScanResults: ScanResults = (resolvedBoxes, Seq.empty, Seq.empty)
 
-    val scanRes = transactions.foldLeft((initialScanResults, previousBoxIds)) { case ((scanResults, accBoxIds), tx) =>
-      val txInputIds = tx.inputs.map(x => encodedBoxId(x.boxId))
+    // Wallet unspent outputs, we fetch them only when Bloom filter shows that some outputs may be spent
+    var prevUnspentBoxes: Seq[TrackedBox] = Seq.empty
+
+    val scanRes = transactions.foldLeft(initialScanResults) { case (scanResults, tx) =>
+
+      // extract wallet- (and external scans) related outputs from the transaction
       val myOutputs = extractWalletOutputs(tx, Some(height), walletVars)
 
-      val boxIds: Seq[EncodedBoxId] = accBoxIds ++ myOutputs.map(x => EncodedBoxId @@ x.boxId)
-      val spendingInputIds = txInputIds.filter(x => boxIds.contains(x))
+      // add extracted outputs to the filter
+      myOutputs.foreach { out =>
+        outputsFilter.put(out.box.id)
+      }
+
+      // Check with the Bloom filter that some outputs maybe spent
+      // Bloom filter is a probabilistic data structure, with false-positives may happen
+      // So we need to filter spendingInputs again
+      val spendingInputs = tx.inputs.filter(inp =>
+        outputsFilter.mightContain(inp.boxId)
+      )
+
+      if (myOutputs.nonEmpty || spendingInputs.nonEmpty) {
 
-      if (myOutputs.nonEmpty || spendingInputIds.nonEmpty) {
-        val spentBoxes = spendingInputIds.map { inpId =>
-          registry.getBox(decodedBoxId(inpId))
-            .orElse(scanResults._1.find(tb => tb.box.id.sameElements(decodedBoxId(inpId)))).get //todo: .get
+        val spentBoxes: Seq[TrackedBox] = if (spendingInputs.nonEmpty) {
+
+          // Read unspent boxes before this block from the database
+          if (prevUnspentBoxes.isEmpty) {
+            prevUnspentBoxes = registry.allUnspentBoxes()
+          }
+          // Add unspent boxes from this block
+          val unspentBoxes = prevUnspentBoxes ++ scanResults._1
```

Comment:
This operation together with the `unspentBoxes.find(...)` looks like performance bottleneck.
The `scanResults` are accumulated along transactions, but appended here on every iteration.
Consider using HashMap for faster lookups.

Reply:
done (with mutable.Map for unspentBoxes)